python/cudf/dataframe/multiindex.py

```diff
@@ -25,14 +25,24 @@ class MultiIndex(Index):
     names: Name for each level
     """
 
-    def __init__(self, levels, codes=None, labels=None, names=None):
+    def __init__(self, levels=None,
+                 codes=None, labels=None, names=None, **kwargs):
         self.names = names
         column_names = []
         if labels:
             warnings.warn("the 'labels' keyword is deprecated, use 'codes' "
                           "instead", FutureWarning)
         if labels and not codes:
             codes = labels
+
+        # early termination enables lazy evaluation of codes
+        if kwargs.get('source_data'):
+            self._source_data = kwargs['source_data']
+            self._codes = None
+            self._levels = None
```

Comment:
We should set these to `codes` and `levels` respectively. This way if they're already calculated and we deep copy the MI they're retained.

Reply:
If they aren't calculated it will get the default value of None from the default parameter values.

Reply:
I'm not sure what you mean. I don't think so, if I access .codes or .levels I'll trigger an index generation. If `source_data` is specified to `__init__`, codes and levels are guaranteed to be empty.

Reply:
Since this is in the `init` call, I'm not referring to `self.codes` and `self.levels`, but just `codes` and `levels` that are passed as parameters.

Reply:
oic 👍